compiler/rustc_middle/src/mir/mod.rs

```diff
@@ -1553,7 +1553,18 @@ pub enum StatementKind<'tcx> {
     /// never accessed still get some sanity checks for, e.g., `let x: ! = ..;`
     FakeRead(Box<(FakeReadCause, Place<'tcx>)>),
 
-    /// Write the discriminant for a variant to the enum Place.
+    /// Initialize the place with the given variant and all fields uninit.
```

Comment:
Generators also use `SetDiscriminant` statements, but their usage pattern is different from enums. Fields are effectively present in multiple variants and switching to a different variant should preserve their value.

Reply:
oof. Ok, will do some code reading and come back with ideas

Reply:
It seems like it would suffice for now to restrict the semantics of this to just ADTs. This should still be useful for MIR opts in general. That being said, how does `MaybeLiveLocals` handle this? From reading the code, it looks like it marks the generator as killed on a `SetDiscriminant`

Reply:
On a related note, what prevents generators from being ADTs, at least from the pov of codegen and other things later down the pipeline?